Use Mavo to autosize input boxes & calculate DPI

efeichen opened the pull request:
Same changes as PR #205, but with no conflicts.

efeichen committed:
Updated screens.json

efeichen committed:
Updated screens.json

efeichen committed:
Revert "Updated screens.json"

This reverts commit d37118b52fa8d557357e07ac92fa959fa47bba6b.

efeichen committed:
Updated screens.json

efeichen committed:
Updated screens.json

efeichen committed:
Sync with upstream gh-pages

efeichen committed:
Add mv-autosize class to inputs

efeichen committed:
Add DPI Calculation expressions

efeichen committed:
Delete redundant JS functions & minor formatting

efeichen committed:
Delete screens.json

LeaVerou commented on assets/js/dpi.js:
```diff
@@ -36,16 +36,10 @@ dimension.onchange = update;
 
-$$('#resolutions a, #diagonals a').forEach(function(a) {
-	a.href = '#' + a.textContent;
-});
```
I see you removed this JS, but the functionality is not added via Mavo expressions.
Please try to understand what this code does that you're deleting. It goes over the common resolutions and diagonals and makes them links to setting the corresponding inputs to these values. Links that can be shared. Currently, with your code, clicking them still works, but does not give you a shareable URL. That's very easily fixed by using an expression.

LeaVerou commented on index.html:
```diff
@@ -16,57 +16,65 @@ <h1>
+				<a href="#" tabindex="-1" mv-action="set(width, 1440) set(height, 900)">1440×900</a>
+				<a href="#" tabindex="-1" mv-action="set(width, 1366) set(height, 768)">1366×768</a>
+				<a href="#" tabindex="-1" mv-action="set(width, 1280) set(height, 800)">1280×800</a>
+				<a href="#" tabindex="-1" mv-action="set(width, 1024) set(height, 768)">1024×768</a>
```
So you switched to using Mavo for these, but even though you see it's a very canonical, repetitive set of data, you did not use a collection for it. Instead you painfully copied and pasted the same code multiple times, changing the numbers each time to match the content of the element. Why?

Repeating information is bad practice, because eventually you will need to change the information and you may forget to update it in all the places that duplicate it. It's a huge tenet of software engineering that you should [never repeat yourself](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). This applies to every kind of coding, in every language. Repeating yourself is fine for experimentation (because often repetitive code is easier to author), but should never be the state that code is left in.

Also, not using a collection here is underutilizing Mavo: Why accept contributions for devices and not for these? You can even use the same data file.

Same applies to the diagonals below.

LeaVerou commented on index.html:
```diff
@@ -16,57 +16,65 @@ <h1>
+					if(dimension=d, sqrt(width * width + height * height)/physical,
+					if(dimension=w, width/phyiscal,
+					if(dimension=h, height/physical)))
+				)">
```
This needs better formatting to make it clear what's inside what. At first I thought the conditionals were adjacent, then I noticed the parentheses.

LeaVerou reviewed (changes requested):
Overall good, please see comments and make changes accordingly.

Have you verified that the shape of the display behind the DPI still changes with every calculation like on the current website?

efeichen committed:
Add resolution & diagonal to data file

efeichen committed:
Add minor formatting

efeichen committed:
Change common res and diagonals to Mavo collections & format expression

efeichen committed:
Fix storage link

efeichen committed:
Add additional cleanup

efeichen committed:
Improve formatting for dpi calculation result expression

efeichen committed:
Resolve property & selection value conflict

efeichen commented:
I think I have fixed the things you asked. Thanks for the criticism (especially the part about DRY). I believe every calculation does change the screen's shape since that part still depends on dpi.js.

After I converted the resolutions and diagonals to collections, though, the items in the collections have underlines that are all connected which makes them hard to read (see screenshot). 

<img width="857" alt="screen shot 2018-06-10 at 10 48 37 pm" src="https://user-images.githubusercontent.com/16854657/41210654-ff95d72c-6d00-11e8-9742-1951b11670f7.png">

Is there a way to fix that? Feel free to let me know any additional things to fix, I'm not sure if my result expression formatting is good enough.

efeichen committed:
Add space between res & diagonal anchors

efeichen committed:
Add missing "none" to mv-storage

LeaVerou commented on assets/css/style.css:
```diff
@@ -70,10 +70,18 @@ input#height {
 
+[mv-multiple="resolution"] {
+	margin: 0 2px 0 2px;
+}
```
If you use pixels here, that means if the font size ever changes, you need to adjust your spacing. Remember DRY? It applies to CSS as well. In some cases repetition is unavoidable, but in this case, we can actually express our intent better with code, by using `em` units, which scale nicely with the font size. Always use `em`s if something would benefit by scaling with the text.

LeaVerou commented on index.html:
```diff
@@ -16,57 +16,56 @@ <h1>
+			<input id="physical" property value="13.3" inputmode="numeric" pattern="\d*\.?\d+" class="mv-autosize" mv-storage="none">″
+			<p id="diagonals">Common diagonals:
+				<a mv-multiple="diagonal" href="#[diag]″" tabindex="-1" mv-action="set(physical, diag)" mv-attribute="none">
+					<span property="diag"></span>″
```
Why the nested structure here? (which produces objects in the JSON too) It seems to me that one element would suffice, but I may be missing something.
For resolution above I get it, because there are two pieces of data per item, but here why?

LeaVerou commented on index.html:
```diff
@@ -16,57 +16,56 @@ <h1>
+				<a mv-multiple="resolution" href="#[w]×[h]" tabindex="-1" mv-action="set(width, w) set(height, h)" mv-attribute="none">
+					<span property="w"></span>×<span property="h"></span>
+				</a>
+				<a href="" class="mv-logged-out"><strong class="mv-login">Add</strong></a>
```
I haven't looked at this live, but I suspect the default UI may be a little jarring, we may want to hide `.mv-add-resolution` and `.mv-add-diagonal` in the CSS.

efeichen replied:
What do you mean by "hide `.mv-add-resolution` and `.mv-add-diagonal` in the CSS"? Could you elaborate on that?

LeaVerou replied:
These are classes to target the Add buttons that Mavo generates.

LeaVerou reviewed (changes requested):
Nice work, please see comments!

efeichen committed:
Change diagonal to simple collection

efeichen committed:
Change key (diagonal) of device items to resolve conflict with option value

efeichen committed:
Change res & diagonal collection item margin units to em

efeichen committed:
Change screens ID to devices to be more understandable

efeichen committed:
Combine mv-apps and add further improvements

efeichen commented:
I think there *is* conflict between the `value="d"` of the `<option>` in that select menu and the`property="d"` inside the device collection, because after I changed from 2 to 1 Mavo app, the result of the calculation is always NaN, but after I resolved the conflict by renaming that property of the collection to `diag`, everything worked just fine.

efeichen committed:
Fix mv-storage link (always forget :p )

LeaVerou commented on assets/css/style.css:
```diff
@@ -70,10 +70,18 @@ input#height {
 }
 
+[mv-multiple="resolution"] {
+	margin: 0 .2em 0 .2em;
```
Super minor but you don't need to repeat values here, `0 .2em` would get expanded to `0 .2em 0 .2em` anyway.

From https://developer.mozilla.org/en-US/docs/Web/CSS/margin : 

> - When one value is specified, it applies the same margin to all four sides.
> - When two values are specified, the first margin applies to the top and bottom, the second to the left and right.
> - When three values are specified, the first margin applies to the top, the second to the left and right, the third to the bottom.
> - When four values are specified, the margins apply to the top, right, bottom, and left in that order (clockwise).

(`padding` works the same too)

LeaVerou commented on index.html:
```diff
@@ -8,70 +8,64 @@
+			<select id="dimension" property title="Physical dimension (in inches)">
+				<option selected value="d">Diagonal</option>
+				<option value="wide">Width</option>
+				<option value="high">Height</option>
```
Just noticed, why were these values changed?

efeichen replied:
This happens when the values are the same as property names. 
<img width="782" alt="screen shot 2018-06-12 at 2 45 44 pm" src="https://user-images.githubusercontent.com/16854657/41310419-74106000-6e4f-11e8-95da-5cd822081ad0.png">
Maybe I should change the property names instead for consistency?

LeaVerou replied:
Can you investigate why? You have the inspector, you can select that element and try property references and see what they resolve to.

Looking quickly at your expression, it looks like you are using unquoted text, which is flimsy: If you have properties with those names, these references (e.g. `d`) will be the values of the properties, not the literal text you intended. It's a good practice to always use quotes to avoid surprises (e.g. `'d'`). This is likely related to the issues you're having, and probably also the cause of the weirdness when you use `d` etc as property names.

efeichen committed:
Simplify margin property declaration

efeichen committed:
Change diag back to d

efeichen committed:
Resolve reference issue in result expression
